nova/0001-Passing-disk_profile-as-part-of-user_data.patch

```diff
@@ -0,0 +1,488 @@
++from nova import exception
+ from nova import objects
+ from nova import test
+ from nova.tests.unit import fake_instance
++from nova.tests.unit.virt.ironic import fake_disk_configs as fakes
+ from nova.tests.unit.virt.ironic import utils as ironic_utils
+ from nova.virt.ironic import patcher
+ 
+@@ -143,3 +148,27 @@ class IronicDriverFieldsTestCase(test.NoDBTestCase):
+                          'value': str(preserve), 'op': 'add', }]
+             expected += self._expected_deploy_patch
+             self.assertPatchEqual(expected, patch)
++
++    def test_generic_get_deploy_patch_disk_profile(self):
++        node = ironic_utils.get_test_node(driver='fake')
++        self.instance.user_data = base64.b64encode(fakes.CORRECT_DISK_CONFIG)
++        patch = patcher.create(node).get_deploy_patch(
++                self.instance, self.image_meta, self.flavor)
++        for path in patch:
++            if path["path"] == '/instance_info/ybiip':
++                self.assertEqual(jsonutils.loads(path["value"]),
++                                 yaml.load(fakes.CORRECT_DISK_CONFIG))
++
++    def test_generic_get_deploy_patch_disk_profile_bad_schema(self):
++        node = ironic_utils.get_test_node(driver='fake')
++        self.instance.user_data = base64.b64encode(fakes.INCORRECT_DISK_CONFIG)
++        self.assertRaises(exception.InstanceDeployFailure,
++                          patcher.create(node).get_deploy_patch, self.instance,
++                          self.image_meta, self.flavor)
++
++    def test_generic_get_deploy_patch_disk_profile_bad_yaml(self):
++        node = ironic_utils.get_test_node(driver='fake')
++        self.instance.user_data = base64.b64encode(fakes.BAD_YAML_DISK_CONFIG)
++        self.assertRaises(yaml.YAMLError,
++                          patcher.create(node).get_deploy_patch, self.instance,
++                          self.image_meta, self.flavor)
+diff --git a/nova/virt/ironic/patcher.py b/nova/virt/ironic/patcher.py
+index 651c7012c3..8aba100456 100644
+--- a/nova/virt/ironic/patcher.py
++++ b/nova/virt/ironic/patcher.py
+@@ -17,13 +17,19 @@
+ """
+ Helper classes for Ironic HTTP PATCH creation.
+ """
++import base64
+ 
++import jsonschema
++import logging
+ from oslo_serialization import jsonutils
+-
++import yaml
+ 
+ import nova.conf
++from nova import exception
++from nova.virt.ironic.validation import DISK_CONFIG_SCHEMA
+ 
+ CONF = nova.conf.CONF
++LOG = logging.getLogger(__name__)
+ 
+ 
+ def create(node):
+@@ -105,4 +111,28 @@ class GenericDriverFields(object):
+         if capabilities:
+             patch.append({'path': '/instance_info/capabilities',
+                           'op': 'add', 'value': jsonutils.dumps(capabilities)})
++
++        # add disk_profile to instance_info
++        if instance.user_data:
++            user_data = base64.b64decode(instance.user_data)
++            try:
++                user_data = yaml.load(user_data)
++                if 'disk_config' in user_data:
```

Comment:
I still think we want to prefix this user_data key with `ybiip_` or `yahoo_` or something.